### PR TITLE
User keymap checking for collision

### LINKED ISSUE #1139: Keymaps are not checked for keybinding collision
in registerKeybindings checks are done to make sure the keybinding does not collide.

These are not done in setKeyMap

### PR DESCRIPTION
**Pull request by marechal-p**
Fixes #1139

Also cleaned some noise left from past refactoring.

### CODE REVIEW

File: packages/core/src/browser/keybinding.ts [hunk 1]
```diff
@@ -139,9 +140,9 @@ export class KeybindingRegistry {
         }
     }
 
-    registerKeybindings(...bindings: Keybinding[]): void {
+    registerKeybindings(bindings: Keybinding[], scope?: KeybindingScope): void {
```
Comment (akosyakov) [inline packages/core/src/browser/keybinding.ts]: please leave `registerKeybindings` and `registerKeybinding` as it is to avoid api breaking

introduce protected methods and call them which accept the scope, like `doRegisterKeybindings` and `doRegisterKeybinding `

Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: adding an optional parameter to `registerKeybinding` (taking one binding at once) doesn break the API, should I still create a second one ?

Comment (akosyakov) [reply packages/core/src/browser/keybinding.ts]: but i don't think we need scopes for keybindings, it is keymaps related, programmatic keybindings should go to the default scope always.

Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: I see now, ok !

File: packages/core/src/browser/keybinding.ts [hunk 2]
```diff
@@ -139,9 +143,20 @@ export class KeybindingRegistry {
...
+    protected doRegisterKeybindings(bindings: Keybinding[], scope: KeybindingScope = KeybindingScope.DEFAULT) {
         for (const binding of bindings) {
-            this.registerKeybinding(binding);
+            this.doRegisterKeybinding(binding, scope);
+        }
+    }
+
+    protected doRegisterKeybinding(binding: Keybinding, scope: KeybindingScope = KeybindingScope.DEFAULT) {
+        try {
+            if (!this.isKeybindingAvailable(binding)) { // Throws on parse error
```
Comment (akosyakov) [inline packages/core/src/browser/keybinding.ts]: Should not we take the scope into the account?

Please add tests showing that it is possible to add keybindings to the user keymap which overrides default keybindings and does not fail.

Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Well, I thought the scope was already taken into account, because most of this PR is just a refactoring (so that the binding availability check is shared between functions).

There is a concept of shadowing for a command, but shouldn't it consider keybinding shadowing, instead of command shadowing ??

Inside `isKeybindingShadowed` function:
https://github.com/marechal-p/theia/blob/81f5318bf32d27c03a7c1ddd33b21c4859e75867/packages/core/src/browser/keybinding.ts#L338

Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: To be sure about the use case:

1. Some key sequence (A) is bound to a command (B) by default.
2. User wants to map (A) to (C).
3. (A)->(B) still exists, but (A)->(C) has priority?

Comment (akosyakov) [reply packages/core/src/browser/keybinding.ts]: from the user perspective, i would expect that my keybindings have priority and should overwrite defaults, so yes (A)->(C) has priority

cc @svenefftinge

File: packages/core/src/browser/keybinding.ts [hunk 3]
```diff
@@ -11,6 +11,10 @@ import { KeyCode, KeySequence } from './keys';
 import { ContributionProvider } from '../common/contribution-provider';
 import { ILogger } from "../common/logger";
 import { StatusBarAlignment, StatusBar } from './status-bar/status-bar';
+import { CustomError } from '../common/error';
+import { isOSX } from '../common/os';
+
+class KeybindingNotAvailableError extends CustomError { }
```
Comment (kittaakos) [inline packages/core/src/browser/keybinding.ts]: I think it would be better to do this PR without any `CustomError` classes. That should go into a separate PR.
Why?:
 - I do not think, it should be called as `CustomError`.
 - We have not yet decided whether we go with error codes or not.
 - Why would we need that fine granularity of errors, do you handle them and act differently?

File: packages/core/src/browser/keybinding.ts [hunk 4]
```diff
@@ -150,42 +165,70 @@ export class KeybindingRegistry {
...
+        return true;
+    }
+
+    /**
+     * Converts special `ctrlcmd` modifier back to `ctrl` for non-OSX users in a keybinding string.
+     * (`ctrlcmd` is mapped to the same actual key as `ctrl` under non-OSX users)
+     *
+     * @param keybinding The keybinding string to convert.
+     */
+    getConvertedKeybinding(keybinding: string): string {
```
Comment (kittaakos) [inline packages/core/src/browser/keybinding.ts]: Does it need to be public? Is it really an API method?

File: packages/core/src/browser/keybinding.ts (hunk 2 above)
Comment (svenefftinge) [reply packages/core/src/browser/keybinding.ts]: Yes, of course. The user keybindings override the application keybindings. 
Also, the 'key' for overriding a keybinding should be the key sequence not the command ID.

Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: My thoughts too, so registering KeyBindings do not need to check for collisions in different scopes, but only inside the same scope as theirs right ? In other words cross-scope collisions are not avoided, we rather check at execution time which scope has priority.

File: packages/core/src/browser/keybinding.ts (hunk 3 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: In tests yes, I do compare errors.

I made a second [PR](https://github.com/theia-ide/theia/pull/1536) that regrouped the application errors, I think you also saw the [issue](https://github.com/theia-ide/theia/issues/1535).

Because it was better to be able to not just catch `Error` in the tests I followed @epatpol advice of doing the 2 PRs and rebasing this one on the one with errors.

Of course I was waiting for input on the implementation. I have uncommitted changes on this PR also relative to the use case around cross-scope collisions and key sequence priority. Apparently there was a little problem prior to my refactoring.

Comment (epatpol) [reply packages/core/src/browser/keybinding.ts]: @kittaakos I thought GH would handle better a dependency on another patch like Gerrit (where it would clearly show that a patch depends on another, and what changes are part of this one and not the other) but it seems the only way to make this clear is either submit two PRs with no dependencies on each other, or submit one PR with all the changes. The issue with the first approach is that you can't profit from changes made in one in the other one, and the issue with the second approach is that it's sometimes difficult to make a PR very specific without adding too much stuff that may seem unnecessary around it.

All in all the easiest solution is indeed to remove the Error commit and just focus on keymaps for this one.

File: packages/core/src/browser/keybinding.ts (hunk 2 above)
Comment (akosyakov) [reply packages/core/src/browser/keybinding.ts]: yes, it looks like that it would be the easiest change with the current code

**Comment by marechal-p** [comment]
Ok so the refactoring implied some fixes to go with it.

The collision check only happens on the same scope as the binding being registered.

It is only when fetching the bindings for a certain key sequence that the bindings are iterated over from END scope to DEFAULT scope (top to bottom) and filtering any cross-scope-shadowing based on that.

**Comment by marechal-p** [comment]
Added a test that checks more corner cases.

File: packages/core/src/browser/keybinding.ts [hunk 5]
```diff
@@ -22,7 +23,48 @@ export namespace KeybindingScope {
...
+export interface KeybindingsResult {
+    full: Keybinding[];
+    partial: Keybinding[];
+    shadow: Keybinding[];
+    merge(other: KeybindingsResult): KeybindingsResult;
+    filter(fn: KeybindingsResultFilter): KeybindingsResult;
+
+}
+
+export class KeybindingsResultImpl implements KeybindingsResult {
```
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: I can remove it but I wondered how coupled it was to the Keybinding API, if it would break things on other projects ?
@akosyakov

[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Replacing the interface by a class implementation doesn seem to break anything, tests are OK, so its fine by me.

[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/core/src/browser/keybinding.ts]: It is the implementation detail. I won't even put as the top level class but under `KeybindingRegistry` namespace at the bottom of the file, i.e.
```ts
export namespace KeybindingRegistry {
    export class KeybindingsResult { ... }
}
``` 
the same about `KeybindingsResultFilter`, it is not used anywehere and can be inlined in  `filter` function: `filter(fn: (binding: Keybinding) => boolean): KeybindingsResult`

**Comment by akosyakov** [comment]
@epatpol Could you test it?

File: packages/core/src/browser/keybinding.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Indeed, fixed !

**Comment by epatpol** [comment]
I haven't followed the whole discussion, so sorry if I'm late to the party, but I've just tested master and this PR side by side and there seems to be something different. With this PR, if I remap `quickCommand` with 
```
    {
        "command": "quickCommand",
        "keybinding": "ctrl+shift+f4"
    }
```
The shortcut works as expected, however the original ctrl+shift+p still works. Is that intended? Looking at vscode it seems like remapping `Show All Commands` to `Ctrl + Shift F4` makes the original `Ctrl + Shift + P` do nothing, like overriding it.

I wanted to see if it was always like this or a regression (if that's not the way we want it) and indeed from master it seems like we correctly copy vscode's behavior.

So what's the best way to go here? @akosyakov ?

**Comment by marechal-p** [comment]
Looking at the code it boils down to "what count as a collision ?"

Right now I implemented key-sequence collisions inside a scope. I thought that someone wanted to bind its own key sequences to certain commands, leaving other key bindings to the same command active (the use only adds new ones which have priority).
It also plays well with the current way of managing user key bindings inside a JSON file where we could write things like
```
[
    { command: "thing", keybinding: "ctrl+a" },
    { command: "thing", keybinding: "ctrl+b", context: "something" },
]
```

Technically checking for command name collisions is not that hard with the current implementation, we just need to formaly define the use cases for key binding. Users might want something consistent/useful that matches with their instincts so to speak.

**Comment by akosyakov** [comment]
I don't think that having several keybindings for the same command harms, the more important that use keybindings for the same key sequence and context are taking priority over defaults.

**Comment by epatpol** [comment]
I'm okay with this @akosyakov, I just wanted to make sure that was the use case we wanted to have. I remember discussing it with @hexa00 a while back and I think we went with this approach as we believed that it was about remapping a default keybinding as in replacing it. If a user wanted to add several keybindings for a command and context (custom + default one) he could just add a keymaps.json containing both the custom and default key binding.

**Comment by marechal-p** [comment]
I've just noticed that this PR has been there for quite some while already.

Is it good to merge or is it not useful ? @akosyakov

File: packages/core/src/browser/keybinding.ts [hunk 6]
```diff
@@ -132,60 +131,166 @@ export class KeybindingRegistry {
...
+        return isOSX ? keybinding : keybinding.replace(/\bctrlcmd\b/, 'ctrl');
+    }
+
+    /**
+     * Finds collisions for a binding inside a list of bindings (error-free)
+     *
+     * @param bindings the reference bindings
+     * @param binding the binding to match
+     */
+    protected getKeybindingCollisions(bindings: Keybinding[], binding: Keybinding): KeybindingRegistry.KeybindingsResult {
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: Does it really need to be part of the `KeybindingRegistry` namespace? IMO that's just more difficult to read, and AFAIK it's only used in the keybinding file, if it's not really used outside, just don't export it.

File: packages/core/src/browser/keybinding.ts [hunk 7]
```diff
@@ -279,23 +339,43 @@ export class KeybindingRegistry {
         return result;
     }
 
+    /**
+     * Returns wether a binding is being shadowed or not inside scope
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: s/wether/whether, unless you really mean [a castrated ram](http://www.grammar-monster.com/easily_confused/weather_whether_wether.htm) :)

File: packages/core/src/browser/keybinding.ts [hunk 8]
```diff
@@ -279,23 +339,43 @@ export class KeybindingRegistry {
...
+     *
+     * @param binding binding to look if shadowed
+     * @param scope scope where to look
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: I think the doc is outdated, as it doesn't take the scope into account anymore with your changes.

File: packages/core/src/browser/keybinding.ts [hunk 9]
```diff
@@ -279,23 +339,43 @@ export class KeybindingRegistry {
...
-    isKeybindingShadowed(scope: KeybindingScope, binding: Keybinding): boolean {
-        if (scope >= KeybindingScope.END) {
-            return false;
-        }
-
-        const nextScope = ++scope;
-
-        if (this.getScopedKeybindingsForCommand(nextScope, binding.command).length > 0) {
-            return true;
+    isKeybindingShadowedByAny(binding: Keybinding, startScope: KeybindingScope = KeybindingScope.DEFAULT): boolean {
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: I really like the doc above that says what this function does, but I feel like the name is misleading as I would expect it to check if it's shadowed by any starting from the most specific scope. Either make it more generic but expect the user to check the option, or very specific or more clear. Maybe rename it to `isKeybindingShadowed`... Wait that's used just above... Anyway is this function even used anywhere? If not I would suggest removing it.

File: packages/core/src/browser/keybinding.ts [hunk 10]
```diff
@@ -404,9 +486,12 @@ export class KeybindingRegistry {
         const bindings = this.getKeybindingsForKeySequence(this.keySequence);
 
         if (this.tryKeybindingExecution(bindings.full, event)) {
+
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: Remove unwanted empty spaces?

File: packages/core/src/browser/keybinding.ts [hunk 11]
```diff
@@ -463,3 +541,38 @@ export class KeybindingRegistry {
         }
     }
 }
+
+export namespace KeybindingRegistry {
+    export class KeybindingsResult {
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: As mentionned above, I think it would be better to put that in a separate class, I don't see the need for a namespace here @akosyakov ?

File: packages/keymaps/src/browser/keymaps-service.ts [hunk 12]
```diff
@@ -53,6 +53,7 @@ export class KeymapsService implements Disposable, FrontendApplicationContributi
         this.resourceProvider(keymapsUri).then(resource => {
             this.keymapsResource = resource;
             this.toDispose.push(this.keymapsResource);
+
```
Comment (epatpol) [inline packages/keymaps/src/browser/keymaps-service.ts]: Unwanted space?

File: packages/core/src/browser/keybinding.ts [hunk 13]
```diff
@@ -132,60 +131,166 @@ export class KeybindingRegistry {
...
+                    }
+                    case KeySequence.CompareResult.SHADOW: {
+                        result.shadow.push(registeredBinding);
+                        break;
+                    }
+                    default: {
+                        break;
+                    }
                 }
+            } catch (error) {
```
Comment (epatpol) [inline packages/core/src/browser/keybinding.ts]: Is this try...catch really needed anymore? I don't see which method could throw above.

File: packages/core/src/browser/keybinding.ts [hunk 14]
```diff
@@ -132,60 +131,166 @@ export class KeybindingRegistry {
...
             }
         }
     }
 
+    /**
+     * Register a default keybinding to the registry.
+     *
+     * @param binding
+     */
+    registerKeybinding(binding: Keybinding) {
```
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: That was part of the previous modification but I was advised to not touch the API

File: packages/core/src/browser/keybinding.spec.ts [hunk 15]
```diff
@@ -247,6 +251,57 @@ describe('keybindings', () => {
         expect(bindings.length).to.be.equal(1);
         expect(bindings[0].keybinding).to.be.equal(validKeyBinding);
     });
+
+    it("parsing input key sequence and fetching bindings should only show relevant bindings", () => {
```
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.spec.ts]: The message could be better indeed

File: packages/core/src/browser/keybinding.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (epatpol) [reply packages/core/src/browser/keybinding.ts]: Makes sense, let's leave it at that for now.

File: packages/core/src/browser/keybinding.ts (hunk 13 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: `KeySequence.parse` can throw, even if at this point it shouldn't, the `try/catch` is just to eventually prevent any problem.

File: packages/core/src/browser/keybinding.ts (hunk 6 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: I was advised to do so in the previous review, it was as you described before.

File: packages/core/src/browser/keybinding.ts [hunk 16]
```diff
@@ -132,60 +131,166 @@ export class KeybindingRegistry {
...
+        return true;
+    }
+
+    /**
+     * Converts special `ctrlcmd` modifier back to `ctrl` for non-OSX users in a keybinding string.
+     * (`ctrlcmd` is mapped to the same actual key as `ctrl` under non-OSX users)
+     *
+     * @param keybinding The keybinding string to convert.
+     */
+    protected getPlateformKeybinding(keybinding: string): string {
```
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Done.

File: packages/core/src/browser/keybinding.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Done.

File: packages/core/src/browser/keybinding.ts (hunk 7 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Done.

File: packages/core/src/browser/keybinding.spec.ts [hunk 17]
```diff
@@ -152,9 +153,10 @@ describe('keybindings', () => {
 
     it("should register a correct keybinding, then default back to the original for a wrong one after", () => {
         let keybindings: Keybinding[] = [{
-            command: "test.command",
+            command: TEST_COMMAND.id,
             keybinding: "ctrl+c"
         }];
+
         // Get default binding
         const keystroke = keybindingRegistry.getKeybindingsForCommand('test.command');
```
Comment (gorkem) [inline packages/core/src/browser/keybinding.spec.ts]: Why not use `TEST_COMMAND.id`?

File: packages/core/src/browser/keybinding.spec.ts [hunk 18]
```diff
@@ -169,10 +171,12 @@ describe('keybindings', () => {
...
         // Set invalid binding
         keybindings = [{
-            command: "test.command",
+            command: TEST_COMMAND.id,
             keybinding: "ControlLeft+Invalid"
         }];
+
         keybindingRegistry.setKeymap(KeybindingScope.USER, keybindings);
+
         const defaultBindings = keybindingRegistry.getKeybindingsForCommand('test.command');
```
Comment (gorkem) [inline packages/core/src/browser/keybinding.spec.ts]: TEST_COMMAND.id ?

File: packages/core/src/browser/keybinding.ts [hunk 19]
```diff
@@ -132,60 +131,166 @@ export class KeybindingRegistry {
...
     /**
-     * Register a default keybinding to the registry.
+     * Checks for keySequence collisions in a list of Keybindings
      *
-     * @param binding
+     * @param bindings the keybinding reference list
+     * @param binding the keybinding to test collisions for
      */
-    registerKeybinding(binding: Keybinding) {
+    isKeybindingAvailable(bindings: Keybinding[], binding: Keybinding): boolean {
```
Comment (gorkem) [inline packages/core/src/browser/keybinding.ts]: This probably needs a better name `isKeyBindingAvailable` does not really explain what this function does.

File: packages/core/src/browser/keybinding.spec.ts (hunk 17 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.spec.ts]: I missed it, thanks !

File: packages/core/src/browser/keybinding.ts (hunk 19 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: Well, it checks whether or not a keybinding is available within a given keybinding collection... I would stick to this, do you have an idea for the name ?

Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: @gorkem do you still want this to be changed ?

Comment (svenefftinge) [reply packages/core/src/browser/keybinding.ts]: `containsKeybinding` might be more appropriate.

File: packages/core/src/browser/keybinding.ts (hunk 11 above)
Comment (marechal-p) [reply packages/core/src/browser/keybinding.ts]: I was advised to do so in a previous review.

**Comment by akosyakov** [comment]
yes, minimal changes always appreciated 👍
